File: Implementations/Swift/Sources/AstrolabeProtocol/Negotiation/RuntimeCapability.swift
//
//  RuntimeCapability.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public struct RuntimeCapability: Codable, Equatable, Hashable, Sendable {
    /// Open runtime capability identifier.
    public let rawValue: String

    public init(rawValue: String) throws {
        guard !rawValue.isEmpty, rawValue.count <= 128 else {
            throw RuntimeContractValueError.invalidOpenIdentifier(rawValue)
        }
        self.rawValue = rawValue
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(rawValue: container.decode(String.self))
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    public static let applicationInfo = RuntimeCapability(uncheckedRawValue: "applicationInfo")
    public static let hierarchySnapshot = RuntimeCapability(uncheckedRawValue: "hierarchySnapshot")
    public static let nodeDetail = RuntimeCapability(uncheckedRawValue: "nodeDetail")
    public static let attributePatchDiscovery = RuntimeCapability(uncheckedRawValue: "attributePatchDiscovery")
    public static let attributePatching = RuntimeCapability(uncheckedRawValue: "attributePatching")
    public static let requestCancellation = RuntimeCapability(uncheckedRawValue: "requestCancellation")

    private init(uncheckedRawValue: String) {
        rawValue = uncheckedRawValue
    }
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/negotiation/RuntimeCapability.kt
//
//  RuntimeCapability.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Open capability identifier advertised during handshake. */
@JvmInline
@Serializable
public value class RuntimeCapability(
    /** Raw capability value. */
    public val rawValue: String
) {
    init {
        require(rawValue.isNotEmpty() && rawValue.length <= MAXIMUM_CAPABILITY_LENGTH) {
            "Runtime capability must contain between 1 and 128 characters"
        }
    }

    public companion object {
        /** Runtime can describe the inspected application. */
        public val applicationInfo: RuntimeCapability = RuntimeCapability("applicationInfo")
        /** Runtime can capture hierarchy snapshots. */
        public val hierarchySnapshot: RuntimeCapability = RuntimeCapability("hierarchySnapshot")
        /** Runtime can return node details. */
        public val nodeDetail: RuntimeCapability = RuntimeCapability("nodeDetail")
        /** Runtime can describe patchable attributes. */
        public val attributePatchDiscovery: RuntimeCapability = RuntimeCapability("attributePatchDiscovery")
        /** Runtime can apply temporary attribute patches. */
        public val attributePatching: RuntimeCapability = RuntimeCapability("attributePatching")
        /** Runtime supports request cancellation. */
        public val requestCancellation: RuntimeCapability = RuntimeCapability("requestCancellation")
        /** Runtime can report directed cross-tree node relations. */
        public val uiGraphRelations: RuntimeCapability = RuntimeCapability("uiGraphRelations")
    }
}

private const val MAXIMUM_CAPABILITY_LENGTH: Int = 128
